Make CUDAFuture handle any kind of device type

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57052 Merge CUDAFuture into ivalue::Future
* **#57051 Make CUDAFuture handle any kind of device type**
* #57050 Use only generic helpers in CUDAFuture
* #57049 Introduce generic MultiStreamGuard
* #57048 Simplify CUDAMultiStreamGuard
* #57047 Add recordDataPtrOnStream to DeviceGuardImplInterface
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

lw committed:
Make CUDAFuture handle any kind of device type

Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -34,13 +34,7 @@ def __init__(self, *, devices: Optional[List[Union[int, str, torch.device]]] = N
         """
```
I assume docs for `devices` arg will be added in follow PRs?

lw replied:
I mentioned it in the paragraph but didn't add a specific section for it, let me do that.

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -91,19 +91,48 @@ std::string formatSetOfDevices(
-      impl_(c10::impl::getDeviceGuardImpl(c10::kCUDA)),
-      devices_(sortDevices(std::move(devices))) {
+      impl_(getImplForDevices(devices)),
+      devices_(getSortedIndicesOfDevices(impl_, devices)) {
```
curious, any reason for not just storing devices instead of device indices in this change??

lw replied:
At this stage I wanted to change as few things as possible to keep the delta small and thus help see that this is not changing behaviors. But, yes, I was thinking the same and I'll probably do a follow-up "polishing" PR.

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.cpp:
```diff
@@ -91,19 +91,48 @@ std::string formatSetOfDevices(
   return oss.str();
 }
 
+const c10::impl::DeviceGuardImplInterface* getImplForDevices(
```
Looks like this is the same code with the one in `InlineStreamGuard.h`. Shall we dedup?

mrshenli replied:
NVM, they are slightly difference. One takes devices, and another takes streams.

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture handle any kind of device type"


Make CUDAFuture autodetect the devicetype from its arguments (which thus change from DeviceIndices to full Devices). This in fact transforms CUDAFuture into a AnythingFuture, since it's not tied to CUDA in any way anymore. Having made it fully device-agnostic, we'll merge it into ivalue::Future in the next PR.

Differential Revision: [D28032711](https://our.internmc.facebook.com/intern/diff/D28032711/)

[ghstack-poisoned]
